include/picolibrary/vector.h

```diff
@@ -0,0 +1,1326 @@
+     * \param[in] position The position to insert the values before.
+     * \param[in] initializer_list The initializer list containing the values to insert.
+     *
+     * \return An iterator to the inserted elements if the vector has sufficient capacity
+     *         to store the new elements.
+     * \return picolibrary::Generic_Error::INSUFFICIENT_CAPACITY if the vector does not
+     *         have sufficient capacity to store the new elements.
+     */
+    constexpr auto insert( Const_Iterator position, std::initializer_list<Value> initializer_list ) noexcept
+        -> Result<Iterator, Error_Code>
+    {
+        // #lizard forgives the length
+
+        if ( m_size + initializer_list.size() > N ) {
+            return Generic_Error::INSUFFICIENT_CAPACITY;
+        } // if
+
+        auto element = --end();
+
+        for ( ; element >= position; --element ) {
+            new ( element + initializer_list.size() ) Value{ std::move( *element ) };
+
+            element->~Value();
+        } // for
+
+        auto const inserted_elements = element + 1;
+
+        ::picolibrary::for_each(
+            initializer_list.begin(), initializer_list.end(), [ element ]( auto value ) noexcept {
+                new ( ++element ) Value{ value };
+            } );
+
+        m_size += initializer_list.size();
+
+        return inserted_elements;
+    }
+
+    /**
+     * \brief Emplace an element before the specified position in the vector.
+     *
+     * \tparam Arguments Element construction argument types.
+     *
+     * \param[in] position The position to emplace the new element before.
+     * \param[in] arguments Element construction arguments.
+     *
+     * \return An iterator to the emplaced element if the vector has sufficient capacity
+     *         to store the new element.
+     * \return picolibrary::Generic_Error::INSUFFICIENT_CAPACITY if the vector does not
+     *         have sufficient capacity to insert the new element.
+     */
+    template<typename... Arguments>
+    constexpr auto emplace( Const_Iterator position, Arguments &&... arguments ) noexcept
+        -> Result<Iterator, Error_Code>
+    {
+        if ( m_size + 1 > N ) {
+            return Generic_Error::INSUFFICIENT_CAPACITY;
+        } // if
+
+        auto element = --end();
+
+        for ( ; element >= position; --element ) {
+            new ( element + 1 ) Value{ std::move( *element ) };
+
+            element->~Value();
+        } // for
+
+        auto const inserted_element = element + 1;
+
+        new ( ++element ) Value{ std::forward<Arguments>( arguments )... };
+
+        ++m_size;
+
+        return inserted_element;
+    }
+
+    /**
+     * \brief Remove the element at the specified position in the vector.
+     *
+     * \param[in] position The position of the element to remove.
+     *
+     * \return An iterator to the element following the removed element.
+     */
+    constexpr auto erase( Const_Iterator position ) noexcept -> Iterator
```

Comment:
This function can't be `constexpr` since it uses `reinterpret_cast`.
```suggestion
    auto erase( Const_Iterator position ) noexcept -> Iterator
```